plugins/org.jboss.tools.openshift.reddeer/src/org/jboss/tools/openshift/reddeer/requirement/OpenShiftProjectRequirement.java

```diff
@@ -0,0 +1,134 @@
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import org.apache.commons.lang.StringUtils;
+import org.eclipse.osgi.util.NLS;
+import org.jboss.reddeer.common.logging.Logger;
+import org.jboss.reddeer.common.wait.TimePeriod;
+import org.jboss.reddeer.common.wait.WaitUntil;
+import org.jboss.reddeer.junit.requirement.Requirement;
+import org.jboss.tools.openshift.core.connection.Connection;
+import org.jboss.tools.openshift.reddeer.condition.ProjectExists;
+import org.jboss.tools.openshift.reddeer.requirement.OpenShiftProjectRequirement.RequiredProject;
+import org.jboss.tools.openshift.reddeer.utils.DatastoreOS3;
+
+import com.openshift.restclient.ResourceKind;
+import com.openshift.restclient.model.IProject;
+import com.openshift.restclient.model.project.IProjectRequest;
+
+/**
+ * A requirement that makes sure a given project exists. If it doesnt it will get created
+ * 
+ * @author adietish@redhat.com
+ */
+public class OpenShiftProjectRequirement implements Requirement<RequiredProject> {
+
+	private final static Logger LOGGER = new Logger(OpenShiftProjectRequirement.class);
+
+	private RequiredProject projectSpec;
+
+	private IProject project;
+
+	@Retention(RetentionPolicy.RUNTIME)
+    @Target(ElementType.TYPE)
+    public @interface RequiredProject {
+		/**
+		 * the connection to use when connecting. If nothing is provided
+		 * {@link DatastoreOS3#SERVER} and {@link DatastoreOS33#USERNAME} are
+		 * used.
+		 */
+		String connectionURL() default StringUtils.EMPTY;
+
+		/**
+		 * the project name to use. If nothing is provided
+		 * {@link OpenShiftResources#TEST_PROJECT} is used
+		 */
+		String name() default StringUtils.EMPTY;
+		/**
+		 * the project display name to use. If nothing is provided
+		 * {@link #name()} is used
+		 */
+		String displayName() default StringUtils.EMPTY;
+		/**
+		 * the project description to use. If nothing is provided
+		 * {@link #name()} is used
+		 */
+		String description() default StringUtils.EMPTY;
+	}
+	
+	@Override
+	public boolean canFulfill() {
+		return true;
+	}
+
+	@Override
+	public void fulfill() {
+		String projectName = DatastoreOS3.getValueOrDefault(projectSpec.name(), DatastoreOS3.TEST_PROJECT);
+		Connection connection = ConnectionUtils.getConnectionOrDefault(projectSpec.connectionURL());
+		assertNotNull(NLS.bind("No connection for {0} exists", projectSpec.connectionURL()), connection);
+
+		this.project = getOrCreateProject(projectName, connection);
+		new WaitUntil(
+				new ProjectExists(projectName, connection),
+				TimePeriod.LONG);
+	}
+
+
+	private IProject getOrCreateProject(String name, Connection connection) {
+		IProject project = OpenShiftResourceUtils.getProject(name, connection);
+		if (project == null) {
+			LOGGER.debug(NLS.bind("Project {0} doesnt exist yet in {1}, creating it.", name, connection.getHost()));
+			project = createProject(projectSpec, connection);
```

Comment:
This is not OK. If default values are used ( = usage of requirement without any values in requirement annotation) and project with specified name does not exists, then it in if on line 95 it would continue to condition and on line 97 it fails (com.openshift.restclient.OpenShiftException: Exception trying to POST https://10.1.2.3:8443/oapi/v1/projectrequests response code: 422 ProjectRequest "" is invalid: metadata.name: Required value: name or generateName is required). This is because projectSpec is with no values and in create project its assumed there is value at least for project name